jax/experimental/doubledouble.py

```diff
@@ -0,0 +1,175 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Precision doubling arithmetic transform
+
+Following the approach of Dekker 1971
+(http://csclub.uwaterloo.ca/~pbarfuss/dekker1971.pdf).
+"""
+from typing import Any, Callable, Dict, Sequence
+
+from jax.util import curry
+from jax import core, lax
+import jax.numpy as jnp
+import jax.linear_util as lu
+
+
+class DoublingTracer(core.Tracer):
+  def __init__(self, trace, head, tail):
+    self._trace = trace
+    # TODO(vanderplas): check head/tail have matching shapes & dtypes
+    self.head = head
+    self.tail = tail
+
+  @property
+  def aval(self):
+    return core.raise_to_shaped(core.get_aval(self.head))
+
+  def full_lower(self):
+    return self
+
+
+class DoublingTrace(core.Trace):
+  def pure(self, val: Any):
+    return DoublingTracer(self, val, jnp.zeros(jnp.shape(val), jnp.result_type(val)))
+
+  def lift(self, val: core.Tracer):
+    return DoublingTracer(self, val, jnp.zeros(jnp.shape(val), jnp.result_type(val)))
+
+  def sublift(self, val: DoublingTracer):
+    return DoublingTracer(self, val.head, val.tail)
+
+  def process_primitive(self, primitive, tracers, params):
+    func = doubling_rules.get(primitive, None)
+    if func is None:
+      raise NotImplementedError(f"primitive={primitive}")
+    out = func(*((t.head, t.tail) for t in tracers))
+    return DoublingTracer(self, *out)
+
+
+@lu.transformation
+def doubling_transform(*args):
+  with core.new_master(DoublingTrace) as master:
+    trace = DoublingTrace(master, core.cur_sublevel())
+    in_tracers = [DoublingTracer(trace, hi, lo) for hi, lo in args]
+    outputs = yield in_tracers, {}
+    if isinstance(outputs, Sequence):
+      out_tracers = map(trace.full_raise, outputs)
+      result = [(x.head, x.tail) for x in out_tracers]
+    else:
+      out_tracer = trace.full_raise(outputs)
+      result = (out_tracer.head, out_tracer.tail)
+  yield result
+
+
+@curry
+def doubledouble(f, *args):
+  g = doubling_transform(lu.wrap_init(f))
+  # TODO: flatten pytrees
+  args = ((a, jnp.zeros_like(a)) for a in args)
+  out = g.call_wrapped(*args)
+  if isinstance(out, list):
+    return tuple(o[0] + o[1] for o in out)
+  else:
+    return out[0] + out[1]
+
+
+doubling_rules: Dict[core.Primitive, Callable] = {}
+
+def _mul_const(dtype):
+  _nmant = jnp.finfo(dtype).nmant
+  return jnp.array((2 << (_nmant - _nmant // 2)) + 1, dtype=dtype)
+
+def _abs2(x):
+  x, xx = x
+  return jnp.where(
+    lax.sign(x) == lax.sign(xx),
+    (lax.abs(x), lax.abs(xx)),
+    (lax.abs(x), -lax.abs(xx))
+  )
```

Comment:
This works thanks to broadcasting, but making arrays of length 2 only to immediately split them is a little inefficient.

Reply:
Would you use suggest using two where statements and returning the tuple of results?

Reply:
nvm, I see what you're saying.